### PR TITLE
LLM: add alpaca qlora finetuning example

### PR DESCRIPTION
**Pull request by plusbang**
## Description

- Add alpaca qlora finetuning example with single Arc usage

(According to comparison in https://github.com/analytics-zoo/nano/issues/706#issuecomment-1793989686, use ["q_proj","v_proj","k_proj","o_proj","up_proj","down_proj","gate_proj"] as default target modules)

### 4. How to test?
- [ ] Unit test
- [x] Local test

### CODE REVIEW

**Comment by yangw1234** [comment]
This is the original code https://github.com/tloen/alpaca-lora/blob/main/finetune.py, maybe we should also add their license.

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import torch
+import os
+os.environ["ACCELERATE_USE_IPEX"] = "1"
+os.environ["ACCELERATE_USE_XPU"] = "1"
```
Comment (yangw1234) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py:20]: I think we can delete these two lines.

Comment (shane-huang) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py:20]: I think we should make the changes minimal. Use comment to highlight what we have changed.

**Comment by plusbang** [comment]
> This is the original code https://github.com/tloen/alpaca-lora/blob/main/finetune.py, maybe we should also add their license.

yeah, have added.

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py (hunk 1 above)
Comment (plusbang) [reply python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py]: > I think we should make the changes minimal. Use comment to highlight what we have changed.

I have updated the python script, and add comments to necessary code changes like that https://github.com/intel-analytics/BigDL/pull/9276/files#diff-341eee2e71c40dcf033d6c276621c3f4c2e37021e00a6bb71384f7e0612108b4R52

**Comment by plusbang** [comment]
Please @jason-dai take a look.

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# Alpaca QLoRA Finetuning (experimental support)
+
+This example demonstrates how to finetune a llama2-7b model use BigDL-LLM 4bit optimizations using [Intel GPUs](../README.md).
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:3]: This example ports [Alpaca-LoRA]() to BigDL-LLM QLoRA on [Intel GPU]().

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+## 0. Requirements
+To run this example with BigDL-LLM on Intel GPUs, we have some recommended requirements for your machine, please refer to [here](../../README.md#requirements) for more information.
+
+## Example: Alpaca QLoRA Finetuning
+
+This example is ported from [alpaca-lora](https://github.com/tloen/alpaca-lora/blob/main/finetune.py). Just a few code changes are needed to finetune a llama2-7b model using QLoRA with BigDL-LLM 4bit optimizations on Intel GPUs.
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:10]: to finetune a LLaMA-2 model

Comment (plusbang) [reply python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md]: > to finetune a LLaMA-2 model

Have updated.

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# Alpaca QLoRA Finetuning (experimental support)
+
+This example ports [Alpaca-LoRA](https://github.com/tloen/alpaca-lora/tree/main) to BigDL-LLM QLoRA on [Intel GPUs](../../README.md).
+
+## 0. Requirements
+To run this example with BigDL-LLM on Intel GPUs, we have some recommended requirements for your machine, please refer to [here](../../README.md#requirements) for more information.
+
+## Example: Alpaca QLoRA Finetuning
+
+This example is ported from [Alpaca-LoRA](https://github.com/tloen/alpaca-lora/blob/main/finetune.py). Just a few code changes are needed to finetune a LLaMA-2 model using QLoRA with BigDL-LLM 4bit optimizations on Intel GPUs.
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:10]: remove this

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+### 2. Configures OneAPI environment variables
+```bash
+source /opt/intel/oneapi/setvars.sh
+```
+
+### 3. Finetune model
+
+Before running the example, please make sure you have prepared the [templates folder](https://github.com/tloen/alpaca-lora/tree/main/templates) in the current directory.
+
+Example usage on single Arc:
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:33]: ### 4. Finetuning LLaMA-2-7B on a single Arc:

**Comment by plusbang** [comment]
@jason-dai Now, the minimal necessary components to run Alpaca QloRA are updated, but I'm not sure whether need to add license comments in the JSON files of templates folder.

**Comment by jason-dai** [comment]
> @jason-dai Now, the minimal necessary components to run Alpaca QloRA are updated, but I'm not sure whether need to add license comments in the JSON files of templates folder.

Add comment "This file is copied from ..."

**Comment by plusbang** [comment]
> > @jason-dai Now, the minimal necessary components to run Alpaca QloRA are updated, but I'm not sure whether need to add license comments in the JSON files of templates folder.
> 
> Add comment "This file is copied from ..."

OK, have added comment in JSON files.

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+### 1. Install
+
+```bash
+conda create -n llm python=3.9
+conda activate llm
+# below command will install intel_extension_for_pytorch==2.0.110+xpu as default
+# you can install specific ipex/torch version for your need
+pip install --pre --upgrade bigdl-llm[xpu] -f https://developer.intel.com/ipex-whl-stable-xpu
+pip install transformers==4.34.0
+pip install fire datasets peft==0.5.0
```
Comment (rnwang04) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:17]: add `pip install accelerate==0.23.0` ?

Comment (plusbang) [reply python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:17]: > add `pip install accelerate==0.23.0` ?

Following the provided commands, `accelerate==0.21.0` is installed and experiments work fine. Is it necessary?

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/templates/alpaca.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/templates/alpaca.json:1]: add comment for where this file is copied from

Comment (plusbang) [reply python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/templates/alpaca.json:1]: > add comment for where this file is copied from

Have updated.

**Review by jason-dai — APPROVED** [review]
LGTM

**Pull request merged**